cluster-autoscaler/estimator/binpacking_estimator.go

```diff
@@ -0,0 +1,124 @@
+type BinpackingNodeEstimator struct {
+	predicateChecker *simulator.PredicateChecker
+}
+
+// NewBasicNodeEstimator builds BasicNodeEstimator.
+func NewBinpackingNodeEstimator(predicateChecker *simulator.PredicateChecker) *BinpackingNodeEstimator {
+	return &BinpackingNodeEstimator{
+		predicateChecker: predicateChecker,
+	}
+}
+
+// Estimate implements First Fit Decreasing bin-packing approximation algorithm.
+// See https://en.wikipedia.org/wiki/Bin_packing_problem for more details.
+// While it is a multi-dimensional bin packing (cpu, mem, ports) in most cases the main dimension
+// will be cpu thus the estimated overprovisioning of 11/9 * optimal + 6/9 should be
+// still be maintained.
+// It is assumed that all pods from the given list can fit to nodeTemplate.
+// Returns the number of nodes needed to accommodate all pods from the list.
+func (estimator *BinpackingNodeEstimator) Estimate(pods []*kube_api.Pod, nodeTemplate *schedulercache.NodeInfo) int {
+
+	podInfos := calculatePodScore(pods, nodeTemplate)
+	sort.Sort(byScoreDesc(podInfos))
+
+	// nodeWithPod function returns NodeInfo, which is a copy of nodeInfo argument with an additional pod scheduled on it.
+	nodeWithPod := func(nodeInfo *schedulercache.NodeInfo, pod *kube_api.Pod) *schedulercache.NodeInfo {
+		podsOnNode := nodeInfo.Pods()
+		podsOnNode = append(podsOnNode, pod)
+		newNodeInfo := schedulercache.NewNodeInfo(podsOnNode...)
+		newNodeInfo.SetNode(nodeInfo.Node())
+		return newNodeInfo
+	}
+
+	nodeTemplate.Node().Status.Allocatable = nodeTemplate.Node().Status.Capacity
+
+	newNodes := make([]*schedulercache.NodeInfo, 0)
+	for _, podInfo := range podInfos {
+		found := false
+		for i, nodeInfo := range newNodes {
+			if err := estimator.predicateChecker.CheckPredicates(podInfo.pod, nodeInfo); err == nil {
+				found = true
+				newNodes[i] = nodeWithPod(nodeInfo, podInfo.pod)
+				break
+			}
+		}
+		if !found {
+			newNodes = append(newNodes, nodeWithPod(nodeTemplate, podInfo.pod))
+		}
+	}
+	return len(newNodes)
+}
+
+// Calculates score for all pods and returns podInfo structure.
+// Score is defined as cpu_sum/node_capacity + mem_sum/node_capacity.
+// Pods that have bigger requirements should be processed first, thus have higher scores.
+func calculatePodScore(pods []*kube_api.Pod, nodeTemplate *schedulercache.NodeInfo) []*podInfo {
+	podInfos := make([]*podInfo, 0, len(pods))
+
+	for _, pod := range pods {
+		cpuSum := resource.Quantity{}
+		memorySum := resource.Quantity{}
+
+		for _, container := range pod.Spec.Containers {
+			if request, ok := container.Resources.Requests[kube_api.ResourceCPU]; ok {
+				cpuSum.Add(request)
+			}
+			if request, ok := container.Resources.Requests[kube_api.ResourceMemory]; ok {
+				memorySum.Add(request)
+			}
+		}
+		score := float64(0)
+		if cpuCapacity, ok := nodeTemplate.Node().Status.Capacity[kube_api.ResourceCPU]; ok && cpuCapacity.MilliValue() > 0 {
```

Comment:
Should this use `Allocatable` instead of `Capacity`? You're using `Allocatable` above.

The same below.